### PR TITLE
cardano-cli: simplifiy the consumption of 'InputTxBodyOrTxFile`

### PR DESCRIPTION
**Pull request by MarcFontaine**
This PR makes is easier to use CLI arguments of type `InputTxBodyOrTxFile`
* Reorder `IncompleteTx`
* Add `TxOrTxBody`
* Add `readInputTxBodyOrTxFile :: InputTxBodyOrTxFile -> IO (Either CddlError (InAnyCardanoEra CompleteOrIncompleteTx)`

### CODE REVIEW

**Comment by MarcFontaine** [comment]
This PR implements:
https://github.com/input-output-hk/cardano-node/issues/4646
and
https://github.com/input-output-hk/cardano-node/issues/4576

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 1]
```diff
@@ -818,8 +818,25 @@ pTransaction =
...
+    TxView
+      <$> pInputTxOrTxBodyFile
+      <*> pTxViewFormat
+      <*> pMaybeOutputFile
+
+  pTxViewFormat :: Parser TxViewFormat
+  pTxViewFormat =
+      Opt.flag' TxViewLegacyYAML
+        (  Opt.long "legacy-yaml"
+        <> Opt.help "Use leagay YAML format (default)."
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:831]: Is `leagay` a typo?

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs [hunk 2]
```diff
@@ -482,6 +531,16 @@ instance Error CddlError where
...
   displayError (CddlIOError e) = displayError e
 
+liftAnyCardanoEra :: (forall era. IsCardanoEra era => thing1 era -> thing2 era) -> InAnyCardanoEra thing1 -> InAnyCardanoEra thing2
+liftAnyCardanoEra f (InAnyCardanoEra era a) = InAnyCardanoEra era $ f a
+
+renderCddlError :: CddlError -> Text
+renderCddlError (CddlErrorTextEnv textEnvErr cddlErr) =
+  "Failed to decode neither the cli's serialisation format nor the ledger's \
+  \CDDL serialisation format. TextEnvelope error: " <> Text.pack (displayError textEnvErr) <> "\n" <>
+  "TextEnvelopeCddl error: " <> Text.pack (displayError cddlErr)
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:541]: Suggest avoid multiline string literals and use `<>` instead.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs [hunk 3]
```diff
@@ -1092,42 +1050,26 @@ runTxCalculateMinFee (TxBodyFile txbodyFile) nw protocolParamsSourceSpec
...
-    unwitnessed <- firstExceptT ShelleyTxCmdCddlError . newExceptT
-                     $ readFileTxBody txbodyFile
-    pparams <- firstExceptT ShelleyTxCmdProtocolParamsError
+  pparams <- firstExceptT ShelleyTxCmdProtocolParamsError
                  $ readProtocolParametersSourceSpec protocolParamsSourceSpec
-    case unwitnessed of
-      IncompleteCddlFormattedTx anyTx -> do
-        InAnyShelleyBasedEra _era unwitTx <-
+  anyTx <- readFileTxBody txbodyFile `rescue` ShelleyTxCmdCddlError
+  --TODO: in principle we should be able to support Byron era txs too
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:1056]: Can you say what currently happens with Byron era transactions?

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs [hunk 4]
```diff
@@ -1281,98 +1222,53 @@ runTxCreateWitness
...
+                              witnessFile
+          Just Refl -> return witness
+
+    writeOutput :: IsCardanoEra era => TxFormat -> Tx era -> ExceptT ShelleyTxCmdError IO ()
+    writeOutput fmt tx = case fmt of
+      CddlFormat -> writeTxFileTextEnvelopeCddl oFp tx `rescue` ShelleyTxCmdWriteFileError
+      TextFormat -> writeFileTextEnvelope oFp Nothing tx `rescue` ShelleyTxCmdWriteFileError
+
+rescue :: Functor m =>  m (Either x a) -> (x -> y) -> ExceptT y m a
+rescue action except = firstExceptT except $ newExceptT action
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:1271]: We should probably move this to a re-usable location, but not right now.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs [hunk 5]
```diff
@@ -1281,98 +1222,53 @@ runTxCreateWitness
...
-                                         witnessFile
-                   Just Refl -> return witness
-            | witnessFile@(WitnessFile file) <- witnessFiles ]
-
-        let tx = makeSignedTransaction witnesses txbody
-
-        firstExceptT ShelleyTxCmdWriteFileError . newExceptT $
-          writeTxFileTextEnvelopeCddl oFp tx
+runTxSignWitness inputTx witnessFiles (OutputFile oFp) = do
+  (InAnyCardanoEra era input) <- readInputTxBodyOrTxFile inputTx `rescue` ShelleyTxCmdCddlError
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:1246]: I'm in favour of dropping the redundant parens in this bind.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs [hunk 6]
```diff
@@ -1233,42 +1175,41 @@ runTxHashScriptData scriptDataOrFile = do
...
+    Nothing -> liftIO $ BS.putStr bs
+    Just fp -> writeOutput fp bs `rescue` ShelleyTxCmdWriteFileError
+  where
+    writeOutput (OutputFile path) bs
+      = runExceptT $ handleIOExceptT (FileIOError path) $ BS.writeFile path bs
+    legacyYamlFormat era = \case
+      CompleteTx tx -> friendlyTxBS era tx
+      IncompleteTx (UnwitnessedCliFormattedTxBody txBody) -> friendlyTxBodyBS era txBody
+      -- In the IncompleteCddlFormattedTx case, the legacy format
+      -- takes the body of the tx  and converts the tx to a tx-body
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:1204]: Remove double space after `tx`.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs [hunk 7]
```diff
@@ -1233,42 +1175,41 @@ runTxHashScriptData scriptDataOrFile = do
...
+      -- and prints that using `friendlyTxBodyBS era body`.
+      -- Any preexisting witnesses (that may be there ?!? ) are discarded !!
+      -- Would make more sense:
+      -- IncompleteTx (IncompleteCddlFormattedTx tx) -> friendlyTxBS era tx
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:1208]: This is an interesting point.  @Jimbo4350

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs [hunk 8]
```diff
@@ -458,16 +467,56 @@ readFileTx fp = do
...
-  = UnwitnessedCliFormattedTxBody (InAnyCardanoEra TxBody)
-  | IncompleteCddlFormattedTx (InAnyCardanoEra Tx)
-
-readFileTxBody :: FilePath -> IO (Either CddlError IncompleteTx)
-readFileTxBody fp = do
-  eTxBody <- readFileInAnyCardanoEra AsTxBody fp
-  case eTxBody of
-    Left e -> fmap (IncompleteCddlFormattedTx . unCddlTx) <$> acceptTxCDDLSerialisation e
-    Right txBody -> return $ Right $ UnwitnessedCliFormattedTxBody txBody
+data IncompleteTx era
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:470]: What is the benefit of adding the `era` parameter? This has further complicated the code without any obvious benefit.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
I am highly skeptical of the changes to `data IncompleteTx`. I don't see any obvious benefit for the increase in code complexity.

**Comment by Jimbo4350** [comment]
> The old definition was the root-cause for the code-duplication in `Cardano.CLI.Shelley.Run.Transaction`.

Lets hold off on the refactor for now as:
```
data IncompleteTx
  = UnwitnessedCliFormattedTxBody (InAnyCardanoEra TxBody)
  | IncompleteCddlFormattedTx (InAnyCardanoEra Tx)
```
Will be eliminated when we remove the cli format, which I will like to do in the next cli release. Once we do that we can revisit the refactor.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
https://github.com/input-output-hk/cardano-node/pull/4647#issuecomment-1325201697

**Pull request closed without merging**